File: detect.py
import joblib
import sys
import os
import time
from collections import Counter

MODEL_FILE = "detect.model"
NGRAM_SIZE = 3
AMPLIFICATION_POWER = 8
def get_ngrams(text):
    text = "".join(text.split())
    return [text[i:i+NGRAM_SIZE] for i in range(len(text)-NGRAM_SIZE+1)]

def calculate_similarity(target_dist, profile_dist):
    overlap = 0
    for ngram, freq in target_dist.items():
        if ngram in profile_dist:
            overlap += min(freq, profile_dist[ngram])
    return overlap

def detect_math(content, profiles):
    target_ngrams = get_ngrams(content[:20000])
    if not target_ngrams: return []
    
    counts = Counter(target_ngrams)
    total = sum(counts.values())
    target_dist = {ngram: (count / total) * 100 for ngram, count in counts.items()}

    results = []
    for category, dist in profiles.items():
        score = calculate_similarity(target_dist, dist)
        results.append({"name": category, "prob": score ** AMPLIFICATION_POWER})

    total_prob = sum(r["prob"] for r in results)
    if total_prob > 0:
        for r in results:
            r["confidence"] = (r["prob"] / total_prob) * 100
    else:
        for r in results: r["confidence"] = 0

    results.sort(key=lambda x: x["confidence"], reverse=True)
    return results

def main():
    if len(sys.argv) < 2:
        print("Usage: python detect.py <file.lua>")
        return

    target_file = sys.argv[1]
    if not os.path.exists(target_file):
        print(f"Error: File {target_file} not found.")
        return

    if not os.path.exists(MODEL_FILE):
        print("Error: model not found please download from github")
        return

    print(f"[*] LOADING MATHEMATICAL ENGINE...")
    profiles = joblib.load(MODEL_FILE)

    print(f"[*] ANALYZING: {os.path.basename(target_file)}")
    start_time = time.time()

    with open(target_file, "r", encoding="utf-8", errors="ignore") as f:
        content = f.read()

    results = detect_math(content, profiles)
    end_time = time.time()
    
    if not results:
        print("[!] no script or no match.")
        return

    top = results[0]
    print("\n" + "="*45)
    print(f"PRIMARY DETECTION: {top['name'].upper()}")
    print(f"CONFIDENCE:        {top['confidence']:.2f}%")
    print("="*45)

    print("\n[INFO]")
    for r in results:
        bar_len = int(r["confidence"] / 2.5)
        bar = "#" * bar_len
        print(f"{r['name']:<15} | {bar:<40} | {r['confidence']:>6.2f}%")

    print(f"\nSeconds took: {(end_time - start_time) * 1000:.2f}ms")

if __name__ == "__main__":
    main()

File: detect.js
const fs = require('fs');
const path = require('path');

const MODEL_FILE = 'model.json';
const NGRAM_SIZE = 3;
const AMPLIFICATION_POWER = 8;

function getNGrams(text) {
    const cleanText = text.replace(/\s+/g, '');
    const ngrams = [];
    for (let i = 0; i <= cleanText.length - NGRAM_SIZE; i++) {
        ngrams.push(cleanText.substring(i, i + NGRAM_SIZE));
    }
    return ngrams;
}

function calculateSimilarity(targetDist, profileDist) {
    let overlap = 0;
    for (const [ngram, freq] of Object.entries(targetDist)) {
        if (profileDist[ngram]) {
            overlap += Math.min(freq, profileDist[ngram]);
        }
    }
    return overlap;
}

function detectMath(content, profiles) {
    const targetNGrams = getNGrams(content.slice(0, 20000));
    if (targetNGrams.length === 0) return [];

    const counts = {};
    targetNGrams.forEach(ngram => {
        counts[ngram] = (counts[ngram] || 0) + 1;
    });

    const total = targetNGrams.length;
    const targetDist = {};
    for (const [ngram, count] of Object.entries(counts)) {
        targetDist[ngram] = (count / total) * 100;
    }

    const results = [];
    for (const [category, dist] of Object.entries(profiles)) {
        const score = calculateSimilarity(targetDist, dist);
        results.push({ name: category, prob: Math.pow(score, AMPLIFICATION_POWER) });
    }

    const totalProb = results.reduce((sum, r) => sum + r.prob, 0);
    results.forEach(r => {
        r.confidence = totalProb > 0 ? (r.prob / totalProb) * 100 : 0;
    });

    results.sort((a, b) => b.confidence - a.confidence);
    return results;
}

function main() {
    const args = process.argv.slice(2);
    if (args.length < 1) {
        console.log("Usage: node detect.js <file.lua>");
        return;
    }

    const targetFile = args[0];
    if (!fs.existsSync(targetFile)) {
        console.log(`Error: File ${targetFile} not found.`);
        return;
    }

    if (!fs.existsSync(MODEL_FILE)) {
        console.log("Error: model.json not found. Please run convert_model.py first.");
        return;
    }

    const profiles = JSON.parse(fs.readFileSync(MODEL_FILE, 'utf8'));

    let content;
    try {
        content = fs.readFileSync(targetFile, 'utf8');
    } catch (err) {
        content = fs.readFileSync(targetFile, 'binary');
    }

    const startTime = Date.now();
    const results = detectMath(content, profiles);
    const endTime = Date.now();

    if (results.length === 0) {
        console.log("[!] no script or no match.");
        return;
    }

    const top = results[0];
    console.log(`\n[+] DETECTION: ${top.name.toUpperCase()}`);
    console.log(`[+] CONFIDENCE: ${top.confidence.toFixed(2)}%`);

    console.log("\n[INFO]");
    results.forEach(r => {
        console.log(`${r.name.padEnd(15)} | ${r.confidence.toFixed(1)}`);
    });

    console.log(`\nSeconds took: ${(endTime - startTime).toFixed(2)}ms`);
    console.log("\n[!] if this was wrong please report to xss.in on discord");
}

main();
